### PR TITLE
Add Github Wiki add on for Web variant

### PR DESCRIPTION
**Pull request by bterone**
Closes https://github.com/nimblehq/elixir-templates/issues/102

## What happened

✅: Added the Github Wiki addon to the template
✅. Add `Home.md, Getting-Started.md and _Sidebar.md` when using addon

✔️. Use the new github wiki script using rsync
 
## Insight

Followed a similar structure as the rest of the Github module 😄 

Using the information in `Getting-Started.md.eex` from the `README.md.eex`. There are two versions of the README, one for Phoneix projects, and another for mix projects.

## Proof Of Work

Can add the Github Wiki workflow 🎉 
Screen Shot 2021-12-21 at 4 43 33 PM

Screen Shot 2021-12-21 at 4 43 41 PM

### CODE REVIEW

**Comment by andyduong1920** [comment]
can we have this addon for API, Mix variant too @bterone :)

**Comment by andyduong1920** [comment]
Please rebase this PR @bterone  the test was fixed on #128 👍

But I think we are moving to the new Wiki workflow tool, right @bterone should we consider to implement that new tool to the temaplate

**Comment by bterone** [comment]
Will need to update to use the new Github Wiki script 👍

File: priv/templates/nimble_template/.github/workflows/publish_wiki.yml.eex [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ push:
+    paths:
+      - .github/wiki/**
+    branches:
+      - develop
+
+env:
+  USER_TOKEN: ${{ secrets.GH_TOKEN }}
+  USER_NAME: github-wiki-workflow
+  USER_EMAIL: dev@nimblehq.co
```
Comment (andyduong1920) [inline priv/templates/nimble_template/.github/workflows/publish_wiki.yml.eex:13]: Should we hardcode this as `dev@nimblehq.co` @bterone 🤔 as the template is for the community, how about we change it to something general 🤔 Eg: `EMAIL_ADDESS`

File: lib/nimble_template/addons/github.ex [hunk 2]
```diff
@@ -51,4 +51,14 @@ defmodule NimbleTemplate.Addons.Github do
 
     project
   end
+
+  @impl true
+  def do_apply(%Project{} = project, opts) when is_map_key(opts, :github_wiki) do
+    template_file_path = ".github/workflows/publish_wiki.yml.eex"
+    file_path = ".github/workflows/publish_wiki.yml"
+
+    Generator.copy_file([{:text, template_file_path, file_path}])
```
Comment (andyduong1920) [inline lib/nimble_template/addons/github.ex:60]: should we create the `.github/wiki` folder, along with some initial files like 

- Home
- _Sidebar
- Getting-Started

File: priv/templates/nimble_template/.github/workflows/publish_wiki.yml.eex (hunk 1 above)
Comment (bterone) [reply priv/templates/nimble_template/.github/workflows/publish_wiki.yml.eex:13]: I'm thinking we can keep it consistent with `GH_TOKEN` and call it `GH_EMAIL`, what do you think? @andyduong1920 🤔

Comment (andyduong1920) [reply priv/templates/nimble_template/.github/workflows/publish_wiki.yml.eex:13]: That works @bterone

Comment (bterone) [reply priv/templates/nimble_template/.github/workflows/publish_wiki.yml.eex]: Adjusted in 07e34db910664483d24974a75e9a79f0951e6975 👍

File: lib/nimble_template/addons/github.ex (hunk 2 above)
Comment (bterone) [reply lib/nimble_template/addons/github.ex]: Added in a111487b059b19f8586a303a42d561d67c1d168e

**Review by andyduong1920 — APPROVED** [review]

File: lib/nimble_template/helpers/github.ex [hunk 3]
```diff
@@ -9,4 +9,10 @@ defmodule NimbleTemplate.GithubHelper do
 
   def generate_github_action?(),
     do: Mix.shell().yes?("\nDo you want to generate the Github Action workflow?")
+
+  def generate_github_wiki?(),
+    do:
+      Mix.shell().yes?(
+        "\nDo you want to publish a Github Wiki for this project? You'd need to manually create the first Github Wiki Page and set the GH_TOKEN and GH_EMAIL secret for this properly function."
```
Comment (rosle) [inline lib/nimble_template/helpers/github.ex:16]: Not sure, should it be

```suggestion
        "\nDo you want to publish a Github Wiki for this project? You'd need to manually create the first Github Wiki Page and set the GH_TOKEN and GH_EMAIL secret for this to properly function."
```

File: lib/nimble_template/addons/github.ex [hunk 4]
```diff
@@ -51,4 +51,49 @@ defmodule NimbleTemplate.Addons.Github do
...
+
+    homepage_path = ".github/wiki/Home.md"
+    sidebar_path = ".github/wiki/_Sidebar.md"
+
+    files = [
+      {:text, template_workflow_path, workflow_path},
+      {:text, homepage_path, homepage_path},
+      {:eex, template_getting_started_path, getting_started_path},
+      {:text, sidebar_path, sidebar_path}
+    ]
```
Comment (rosle) [inline lib/nimble_template/addons/github.ex:93]: I might prefer putting some variables that are not reused directly.  

```suggestion
    files = [
      {:text, ".github/workflows/publish_wiki.yml.eex", ".github/workflows/publish_wiki.yml"},
      {:text, homepage_path, homepage_path},
      {:eex, template_getting_started_path, ".github/wiki/Getting-Started.md"},
      {:text, sidebar_path, sidebar_path}
    ]
```

I think it's easier to see what is being copied. What do you think? 💭

File: test/nimble_template/addons/github_test.exs [hunk 5]
```diff
@@ -123,4 +123,132 @@ defmodule NimbleTemplate.Addons.GithubTest do
...
+      end)
+    end
+  end
+
+  describe "#apply/2 with api_project and github_wiki option" do
+    test "copies the .github/workflows/publish_wiki.yml and Github Wiki files", %{
+      project: project,
+      test_project_path: test_project_path
+    } do
+      in_test_project(test_project_path, fn ->
```
Comment (rosle) [inline test/nimble_template/addons/github_test.exs:174]: Do we need to do:  
```
project = %{project | api_project?: true, web_project?: false}
```

Maybe later, we might want to create a helper like `@describetag mix_project?: true` to setup api project for testing to make it more convenient. 🤔 💡

File: priv/templates/nimble_template/.github/workflows/publish_wiki.yml.eex [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      - develop
+
+env:
+  USER_TOKEN: ${{ secrets.GH_TOKEN }}
+  USER_NAME: github-wiki-workflow
+  USER_EMAIL: ${{ secrets.GH_EMAIL }}
+  OWNER: ${{ github.event.repository.owner.name }}
+  REPOSITORY_NAME: ${{ github.event.repository.name }}
+  WIKI_FOLDER: .github/wiki
+  TEMP_CLONE_WIKI_FOLDER: tmp_wiki
```
Comment (rosle) [inline priv/templates/nimble_template/.github/workflows/publish_wiki.yml.eex:17]: I've been thinking about this. Should we change this to `TMP_CLONE_WIKI_FOLDER` 💭

File: lib/nimble_template/addons/github.ex (hunk 4 above)
Comment (bterone) [reply lib/nimble_template/addons/github.ex:93]: Yeah, I'm fine with this approach too 👍

File: priv/templates/nimble_template/.github/workflows/publish_wiki.yml.eex (hunk 6 above)
Comment (bterone) [reply priv/templates/nimble_template/.github/workflows/publish_wiki.yml.eex:17]: To be honest, I'd rather keep it as `TEMP_CLONE_WIKI_FOLDER ` as its clearer as a variable name 🤔

**Review by longnd — APPROVED** [review]

File: lib/nimble_template/addons/github.ex [hunk 7]
```diff
@@ -51,4 +51,44 @@ defmodule NimbleTemplate.Addons.Github do
...
+        ".github/wiki/Getting-Started.md.mix.eex"
+      else
+        ".github/wiki/Getting-Started.md.eex"
+      end
+
+    homepage_path = ".github/wiki/Home.md"
+    sidebar_path = ".github/wiki/_Sidebar.md"
+
+    files = [
+      {:text, ".github/workflows/publish_wiki.yml.eex", ".github/workflows/publish_wiki.yml"},
```
Comment (rosle) [inline lib/nimble_template/addons/github.ex:84]: This template file can be a normal file (without .eex), right? 🤔  

```suggestion
      workflow_path = ".github/workflows/publish_wiki.yml"

      {:text, workflow_path, workflow_path},
```

**Review by rosle — APPROVED** [review]

File: lib/nimble_template/addons/github.ex (hunk 7 above)
Comment (bterone) [reply lib/nimble_template/addons/github.ex]: Thanks for catching that, I've replaced the template file cfd0fdc28d2985265172c6eb01864e864ec80210 👍

**Pull request merged**